### PR TITLE
4 gpu

### PR DESCRIPTION
**Pull request by smilesun**

### CODE REVIEW

File: domainlab/exp_protocol/run_experiment.py [hunk 1]
```diff
@@ -77,6 +77,7 @@ def run_experiment(
     apply_dict_to_args(args, args_algo_as_task)
     apply_dict_to_args(args, hyperparameters)
     apply_dict_to_args(args, misc, extend=True)
+    args.device = str(param_index % 4)
```
Comment (Car-la-F) [inline domainlab/exp_protocol/run_experiment.py:80]: I'm a little confused now. What does this change?

Does this mean it is now hardcoded to use 4 GPU?

Comment (Car-la-F) [reply domainlab/exp_protocol/run_experiment.py:80]: Oh, I just read the title of the pull request

Comment (smilesun) [reply domainlab/exp_protocol/run_experiment.py:80]: Denbi only has 1 GPU right?

**Comment by Car-la-F** [comment]
@smilesun 
I am a little confused. Is `args.device` used somewhere in the code? I could only find that the algorithms (i.e. diva) do set their device like this 

https://github.com/marrlab/DomainLab/blob/5ff421f009f466c29351a60a4cb217a72850e6c1/domainlab/algos/builder_diva.py#L42

Would it make sense to replace this line with `args.device`

**Comment by smilesun** [comment]
> @smilesun I am a little confused. Is `args.device` used somewhere in the code? I could only find that the algorithms (i.e. diva) do set their device like this
> 
> https://github.com/marrlab/DomainLab/blob/5ff421f009f466c29351a60a4cb217a72850e6c1/domainlab/algos/builder_diva.py#L42
> 
> Would it make sense to replace this line with `args.device`

I make this change a few commits ago

see here: 
https://github.com/marrlab/DomainLab/blob/5ff421f009f466c29351a60a4cb217a72850e6c1/domainlab/utils/utils_cuda.py#L13

Since I suspect snakemake only uses 1 GPU when we set NUM_GPUS to be 2 (Alexej did the experiment)  is that DomainLab always use cuda:0, now the change is with args.device, the user can specify which GPU to use.
In the benchmark, we can use the hyper-parameter index mode the NUM_GPUS to use different GPU devices , does that make sense?

**Comment by Car-la-F** [comment]
@smilesun 
Yes, makes sense. I cannot see this changes in the Files changed. Were they included in another pull request?

**Comment by smilesun** [comment]
see answer from stackoverflow:

https://stackoverflow.com/questions/75997356/how-to-correctly-specify-all-available-gpus-on-a-single-workstation-machine-with/75998269?noredirect=1#comment134041118_75998269

**Pull request merged**